src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,227 @@
+using Microsoft.VisualStudio.ProjectSystem.Logging;
+using Microsoft.VisualStudio.ProjectSystem.References;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    [Export(typeof(IBuildUpToDateCheckProvider))]
+    internal class BuildUpToDateCheck : IBuildUpToDateCheckProvider
+    {
+        private const string DisableFastUpToDateCheckProperty = "DisableFastUpToDateCheck";
+
+        private static string[] KnownOutputGroups = { "Symbols", "Built", "ContentFiles", "Documentation", "LocalizedResourceDlls" };
+
+        private readonly IProjectLockService _projectLockService;
+        private readonly IProjectSystemOptions _projectSystemOptions;
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IProjectLogger _projectLogger;
+        private readonly IProjectItemSchemaService _projectItemsSchema;
+        private readonly Lazy<IFileTimestampCache> _fileTimestampCache;
+
+        [ImportingConstructor]
+        public BuildUpToDateCheck(
+            IProjectLockService projectLockService,
+            IProjectSystemOptions projectSystemOptions,
+            ConfiguredProject configuredProject,
+            IProjectLogger projectLogger,
+            [Import(AllowDefault = true)] IProjectItemSchemaService projectItemSchema,
+            [Import(AllowDefault = true)] Lazy<IFileTimestampCache> fileTimestampCache)
+        {
+            _projectLockService = projectLockService;
+            _projectSystemOptions = projectSystemOptions;
+            _configuredProject = configuredProject;
+            _projectLogger = projectLogger;
+            _projectItemsSchema = projectItemSchema;
+            _fileTimestampCache = fileTimestampCache;
+        }
+
+        public async Task<bool> IsUpToDateAsync(BuildAction buildAction, TextWriter logger, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (buildAction != BuildAction.Build)
+            {
+                return false;
+            }
+
+            using (var access = await _projectLockService.ReadLockAsync())
+            {
+                var project = await access.GetProjectAsync(_configuredProject, cancellationToken);
+
+                void Log(string traceMessage)
+                {
+                    _projectLogger.WriteLine($"FastUpToDate: {traceMessage}");
+                }
+
+                Log($"Starting check for project '{project.FullPath}'.");
+
+                var timestampCache = _fileTimestampCache != null ? _fileTimestampCache.Value.TimestampCache : new Dictionary<string, DateTime>(StringComparer.OrdinalIgnoreCase);
+
+                DateTime GetLatestTimestamp(List<string> paths)
```

Comment:
+1
I agree,  I think those methods which do not depend on any state inside this function, should be the member of this class, instead of local functions.  That is the same for the Log method, it only uses the _projectLogger, which is a member of the class, not anything local to this method.  That is especially true when those methods is valid and has its own logic meaning when it lives outside of this method.

I think they should be moved out to be other private member method of this class.  I totally agree with @srivatsn that this makes this method huge and hard to read.